Fix ansible-test code coverage collection.

mattclay opened the pull request:
##### SUMMARY

The last task in a play should now properly report code coverage.

This change should also eliminate empty coverage files, as well as incomplete coverage files resulting from early worker termination.

##### ISSUE TYPE

Bugfix Pull Request

##### COMPONENT NAME

ansible-test

mattclay committed:
Fix ansible-test code coverage collection.

The last task in a play should now properly report code coverage.

This change should also eliminate empty coverage files, as well as incomplete coverage files resulting from early worker termination.

ci_complete
ci_coverage

relrod commented on lib/ansible/config/base.yml:
```diff
@@ -1828,6 +1828,25 @@ TAGS_SKIP:
+    - The number of seconds to sleep between polling loops when checking Task Queue Manager worker processes to verify they have exited cleanly.
+    - This is for internal use only.
+  env: [{name: ANSIBLE_WORKER_SHUTDOWN_POLL_DELAY}]
+  type: integer
```
We discussed this already, but consider making this a `float` since `time.sleep()` can take a float.

mattclay replied:
Yes, this should be a float. I'll update that.

relrod commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -256,6 +257,15 @@ def cleanup(self):
 
     def _cleanup_processes(self):
         if hasattr(self, '_workers'):
+            for attempts_remaining in range(C.WORKER_SHUTDOWN_POLL_COUNT - 1, -1, -1):
```
This feels odd to me - if I set the poll count to 1, I'd expect it to sleep (at most) once. With the `- 1` here, we enter the loop once, but the only value in the `range` is `0`, so when we get to the `if attempts_remaining` below, `attempts_remaining` is 0 (falsey) so we never sleep.

```suggestion
            for attempts_remaining in range(C.WORKER_SHUTDOWN_POLL_COUNT, -1, -1):
```

mattclay replied:
This is intentional. I was trying to do a few things:

- Make it possible to enable the warning without adding a sleep.
- Avoid sleeping on the final iteration, so that the termination pass happens immediately.

I  agree the behavior could be unintuitive and should be commented. I think the documentation of the config at least accurately describes the behavior.

relrod commented on test/lib/ansible_test/_internal/ansible_util.py:
```diff
@@ -85,6 +85,10 @@ def ansible_environment(args, color=True, ansible_config=None):
+            # give TQM worker processes time to report code coverage results
+            # without this the last task in a play may write no coverage file, an empty file, or an incomplete file
+            ANSIBLE_WORKER_SHUTDOWN_POLL_COUNT='10',
+            ANSIBLE_WORKER_SHUTDOWN_POLL_DELAY='1',
```
In my quick testing, 1 second is wayyyyyy more than we need. I set the poll count to 2 (so we sleep once, see comment above), and can go down to `0.02` without triggering the warning. I was going to suggest `20` and `0.5`, but now I'm thinking something like `100` and `0.1`. Same maximum time as what you had, but trims 0.9s of laggy-ness.

mattclay replied:
I agree this makes more sense for reducing lag, although it may slightly increase CPU usage due to the additional polling. Thanks for verifying it works with a lower sleep.

I'm also thinking that this should always be enabled for testing, not just for coverage. If we're seeing the warning then that's an indication that a worker has hung after reporting results, which is something worth knowing.

mattclay committed:
Code review feedback:

- Change delay config from integer to float.
- Change sleep from 1 second to 0.1 seconds.
- Change polling for tests from 10 to 100 iterations.
- Enable polling for all tests, not just integration tests with coverage.

ci_complete
ci_coverage
